src/core/objects/PDFArray.ts

```diff
@@ -47,6 +47,14 @@ class PDFArray extends PDFObject {
     return this.array[index];
   }
 
+  map(mapFunc: (obj: PDFObject) => PDFObject): PDFArray {
+    const mappedPDFArray = new PDFArray(this.context);
+    this.array.forEach((obj) => {
+      mappedPDFArray.push(mapFunc(obj));
+    });
+    return mappedPDFArray;
+  }
```

Comment:
@cshenks 

I'm a big fan of declarative/functional programming and I'm a strong advocate for using `map`, `filter`,  and `forEach` over imperative `for` and `while` loops. However, when I was building pdf-lib I had to keep performance in mind (PDF processing is very intensive, especially for larger PDF files). So I decided to benchmark various ways of processing arrays in JavaScript: https://jsperf.com/array-testing-2. (see [jsperf_array_tests_chrome.pdf](https://github.com/Hopding/pdf-lib/files/3991917/jsperf_array_tests_chrome.pdf.pdf) and [jsperf_array_tests_firefox.pdf](https://github.com/Hopding/pdf-lib/files/3991918/jsperf_array_tests_firefox.pdf) for the results).

The results are rather unfortunate. `map`, `filter`, and `forEach` are several times slower than their `for` loop equivalents. So I decided to avoid using `map`, `filter`, and `forEach` in the pdf-lib codebase and to stick with raw `for` loops. It's a tough decision to make, because `for` loops are easy to mess up (off-by-one errors, anyone?) and are significantly less readable. However, I think the tradeoff is worth it since pdf-lib is sometimes used to process rather large PDF files.

So, given all of this, I would prefer to avoid introduced a `map` utility here and instead just use a `for` loop when needed. But if there is a common pattern that is used throughout the codebase (e.g. looking up each member of a `PDFArray`), we could introduce a helper on `PDFArray`. For example:
```js
lookupElements(): PDFObject[] {
  const values = new Array(this.size());
  for (let idx = 0, len = this.size(); idx < len; idx++) {
    values[idx] = this.context.lookup(this.get(idx));
  }
  return values;
}
```

Reply:
Updated